src/DataFramesMeta.jl

```diff
@@ -387,18 +386,32 @@ function transform(d::Union{AbstractDataFrame, AbstractDict}; kwargs...)
     return result
 end
 
+
 function transform(g::GroupedDataFrame; kwargs...)
     result = DataFrame(g)
-    idx2 = cumsum(Int[size(g[i],1) for i in 1:length(g)])
-    idx1 = [1; 1 + idx2[1:end-1]]
-    for (k, v) in kwargs
-        first = v(g[1])
-        result[k] = Array{eltype(first)}(size(result, 1))
-        result[idx1[1]:idx2[1], k] = first
-        for i in 2:length(g)
-            result[idx1[i]:idx2[i], k] = v(g[i])
+
+    function spread_scalar(x::Vector, obs_in_group::Int)
+        if length(x) == obs_in_group 
+            return x 
+        else
+            throw(DimensionError("If a function returns a vector, the result " * 
+                                 "must have the same length as the groups it " *
+                                 "operates on"))
         end
     end
+
+    function spread_scalar(x::CategoricalArrays.CategoricalValue, obs_in_group::Int)
+        vec = CategoricalArray(fill(x, obs_in_group))
+        levels!(vec, CategoricalArrays.index(CategoricalArrays.pool(x)))
+    end
+
+    function spread_scalar(x::Any, obs_in_group::Int)
+        fill(x, obs_in_group)
+    end
+
+    for (k, v) in kwargs
+        result[k] = reduce(vcat, spread_scalar(v(ig), size(ig, 1)) for ig in g)
```

Comment:
In theory using a generator as you do is better, but in practice I'm afraid that's going to be slow because the optimized `reduce(vcat, ...)` method from https://github.com/JuliaLang/julia/pull/27188 doesn't exist for generators. For a large number of groups, allocating a new copy for each new entry is going to kill performance.

I've filed https://github.com/JuliaLang/julia/issues/28691 to track this. In the meantime, better keep the existing approach which creates columns and fills them manually, with a reference to that issue.

Reply:
> In the meantime, better keep the existing approach which creates columns and fills them manually, with a reference to that issue.

Thanks. The original problem is that if the first group returns `[1,2,3]` and the second group returns `[missing, missing, missing]`, there is an error because the original vector allocated would be of type `Int`. 

Do you want me to re-work this PR so that `promote_type` works better and we allocate the right vector the first time? 

Or should we put this PR on hold and wait for `vcat` to work better?

Reply:
OK. Then the solution is to do something like `mapreduce(eltype, promote_type, A)` to choose the best element type, as in https://github.com/JuliaLang/julia/pull/27188.

I don't think we can wait for `vcat` to improve, since it's not clear when it will happen (maybe it needs to wait for 2.0 to avoid breakage). See https://github.com/JuliaLang/julia/issues/18472.